### DISCUSSION TITLE
Do we need a `bitcoin-utils` crate?

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2938] [Q&A]
We have `bitcoin-internals` that is supposed to be a perma-non-stable crate and its types should therefore not appear in our public API. We have already at least one violation of this

```bash
bitcoin/src/crypto/key.rs:225:pub struct SortKey(ArrayVec<u8, 65>);
```

and `ArrayVec` comes from `internals`.

Recently:
- While working on `primitives` I found myself wanting public serde modules (currently in `bitcoin::serde_utils`).
- While working on removing `VarInt` I found myself wanting a public `ToU64` trait.

Where should these sorts of things live?
- Arbitrarily in the lowest crate that uses them - implies dumping stuff in `units`
- In a separate `bitcoin-utils` crate (with a different name because that name is taken and the usual problems with "util" as a name)

**Comment by apoelstra** [comment]
`bitcoin_common` is available.

I do think we need such a crate, for serde stuff if nothing else.

**Comment by Kixunil** [comment]
No, that's just lazy thinking again. We have worked so hard to remove `util` module and you want to bring it back? If `ToU64` should be public, it's also universally useful - the crate should not have `bitcoin-` name. But I think I've seen such crate. I though @dpc had one but I don't see it in his repo/crates, maybe he mentioned it somewhere.

> We have already at least one violation of this

I see no violation, that field is private and the struct exists precisely to hide the inner type.

Looking at `bitcoin::serde_utils`, nothing there prevents it from going to `internals`. There are tools to implement proper `serde` public API not the API itself. But looking even deeper there's hex stuff that can go to `hex-conservative` if other people want it, `btreemap_as_seq*` which seems to be PSBT-specific (does PSBT actually have a stable Json serialization definition?!), so it can live there and if anyone needs to use it for something else, they can make a crate. (Also I'm pretty sure I've seen a derive that does this exactly and I've even used it.) There's string stuff that I can't believe doesn't exist anywhere already. (Well, there are attributes but the macro also gives us the ability to specify the expected thing.) We can publish it separately if needed but again, this is completely universal, not bitcoin-specific, so we shouldn't need that.

**Reply by apoelstra** [reply]
The Rust ecosystem does not have the tooling to maintain separate crates for every little thing like `ToU64`. Even `ordered` makes me a little uncomfortable. It's an extra dep without the `bitcoin-` name in it whose purpose is unclear from the name and where I have to constantly remember "that's ours, it's not going to break MSRV or worse".

>does PSBT actually have a stable Json serialization definition

No :P.

**Reply by Kixunil** [reply]
I'd name the crate something like `restrict-architecture`.

> No :P.

Then why do we implement it? Sounds like it'd be better to remove it to avoid confusion.

**Reply by apoelstra** [reply]
Why implement serde de/serialization for PSBT? So that people can easily store them and transmit them over ad-hoc channels.

As for `restrict-architecture` ... this crate would be like 5 lines of code. A couple lines to `compile_error` on the wrong architecture and a few lines to define a `ToU64` trait which we don't even need for interoperability, just to encapsulate some `unwrap`s. (Actually, I wonder if it *can* go into bitcoin-internals for this reason..)

**Reply by dpc** [reply]
@Kixunil https://crates.io/crates/convi ?

I'm a bit confused what is going on here.

**Reply by Kixunil** [reply]
> Why implement serde de/serialization for PSBT? So that people can easily store them and transmit them over ad-hoc channels.

How about we serialize it as base64 encoding of the consensus encoding? Seems more robust, simpler and obvious to me.

> this crate would be like 5 lines of code

So what? If it defines a trait that should be potentially public and global for the ecosystem then it's a good enough justification because of how Rust ecosystem works. It's also reasonably convenient to add a line to `Cargo.toml`. (Potentially, this trait should be in `core` but it's hairy.)

> which we don't even need for interoperability, just to encapsulate some `unwrap`s. (Actually, I wonder if it _can_ go into bitcoin-internals for this reason..)

I only partially agree here. It's supposed to be used in `WriteExt` method signature which makes it part of public API. We can seal it an we can tell people how to use it properly to not break anything but those are just words. If we want it enforced we should either put it into a stable crate or take `u64` in the `WriteExt` method and just convert manually. Given how resistant people are to having a micro-crate I'm now in favor of the latter approach. Edit: though I just realized it is a bit annoying for downstream users to have to define their own even if the validity of conversion is implied by `bitcoin` not supporting 128-bit.

@dpc yes, that one, thanks!

**Reply by apoelstra** [reply]
>How about we serialize it as base64 encoding of the consensus encoding? Seems more robust, simpler and obvious to me.

Yeah, we should do this. May take a bit of work to make sure it's possible without allocations (though this isn't urgent; `bitcoin` already depends on having an allocator so this is purely a performance thing).

**Reply by apoelstra** [reply]
[comment deleted; converted to issue #2941 since Github is hiding messages here and it was getting long and far-afield anyway]

**Comment by tcharding** [comment]
> I see no violation, that field is private and the struct exists precisely to hide the inner type.

You are correct, my bad. No idea what I was grepping or thinking when I wrote that.

> But looking even deeper there's hex stuff that can go to `hex-conservative` if other people want it

Done last week already in https://github.com/rust-bitcoin/hex-conservative/pull/92

Why don't I try to update #2931  using `convi` and see how it looks?

**Reply by Kixunil** [reply]
Either that (but we'd have to collaborate with @dpc to stabilize `convi`) or remove the trait from the public API.

**Reply by dpc** [reply]
`convi` is <100 lines of non-comment non-test code. I'd be happy to pass the ownership to rust-bitcoin org, or you could maybe inline, fork or whatever else. Again - I'm a bit confused about the details right now, but you have my full cooperation.

**Reply by Kixunil** [reply]
Thanks! We'll ping you if we decide to do it.
